File: bin/03_integrate/dealplus.R
### Date: 250812 dealplus.R
### Image: integration-R-- /opt/conda/bin/R
### Coder: ydgenomics

library(Seurat)
library(optparse)
library(ggplot2)

option_list <- list(
  make_option(c("-i", "--input_rds"),
    type = "character", default = NULL,
    help = "Path to input preprocessed rds file"
  ),
  make_option(c("-o", "--out_rds"),
    type = "character", default = NULL,
    help = "integrated rds file"
  ),
  make_option(c("-p", "--out_UMAP"),
    type = "character", default = NULL,
    help = "Output UMAP after integration"
  ),
  make_option(c("-t", "--other1_key"),
    type = "character", default = NULL,
    help = "Output UMAP after integration"
  ),
  make_option(c("-f", "--other2_key"),
    type = "character", default = NULL,
    help = "Output UMAP after integration"
  )
)

opt <- parse_args(OptionParser(option_list = option_list))
input_rds <- opt$input_rds
out_rds <- opt$out_rds
out_UMAP <- opt$out_UMAP
other1_key <- opt$other1_key
other2_key <- opt$other2_key

obj <- readRDS(input_rds)
#add a column
obj$biosample.celltype <- paste0(obj$biosample, "_", obj$celltype)
unique(obj$biosample.celltype)

#visual
pdf(out_UMAP, width=12, height=8)
DimPlot(obj, reduction = "umap", split.by = "biosample")
DimPlot(obj, reduction = "umap", group.by = "biosample", shuffle = TRUE, label = TRUE)
# DimPlot(obj, reduction = "umap", group.by = "sample", shuffle = TRUE, label = TRUE)
DimPlot(obj, reduction = "umap", group.by = "celltype", shuffle = TRUE, label = TRUE)
DimPlot(obj, reduction = "umap", group.by = "biosample.celltype", shuffle = TRUE, label = TRUE)
# 检查 meta.data 中是否包含指定列
required_features <- c("nCount_RNA", "nFeature_RNA")
if (all(required_features %in% colnames(obj@meta.data))) {
  VlnPlot(obj, features = required_features, group.by = "biosample")
} else {
  message("meta.data lacked nCount_RNA or nFeature_RNA, so passed vioplot")
}

# 检查 meta.data 中是否存在这两个列
if (other1_key %in% colnames(obj@meta.data) && other2_key %in% colnames(obj@meta.data)) {
  # 检查是否满足条件：other1_key 不等于 "biosample" 或 other2_key 不等于 "celltype"
  if (other1_key != "biosample" || other2_key != "celltype") {
    obj$other1_other2 <- paste0(
      obj@meta.data[[other1_key]], "_", obj@meta.data[[other2_key]]
    )
    p1 <- DimPlot(obj, reduction = "umap", split.by = other1_key)
    p2 <- DimPlot(obj, reduction = "umap", group.by = other1_key, shuffle = TRUE, label = TRUE)
    p3 <- DimPlot(obj, reduction = "umap", group.by = other2_key, shuffle = TRUE, label = TRUE)
    p4 <- DimPlot(obj, reduction = "umap", group.by = "other1_other2", shuffle = TRUE, label = TRUE)
    print(p1); print(p2); print(p3); print(p4)
  } else {
    message("other1_key 或 other2_key 为默认值，跳过合并列。")
  }
} else {
  message("meta.data 中缺少指定的列，跳过合并列。")
}
dev.off()
#save dealed rds
saveRDS(obj, file = out_rds)

File: bin/03_integrate/dealplus.py
### Date: 250812 dealplus.py
### Image: integration-R-- /opt/conda/bin/python
### Coder: ydgenomics

import click
import matplotlib.pyplot as plt
import scanpy as sc
from matplotlib.backends.backend_pdf import PdfPages

@click.command()
@click.argument("input_h5ad", type=click.Path(exists=True))
@click.argument("out_h5ad", type=click.Path(exists=False), default=None)
@click.argument("out_umap", type=click.Path(exists=False), default=None)
@click.option('--other1_key', type=str, default=None, help="Sample key to distinguish sample")
@click.option('--other2_key', type=str, default=None, help="Cluster key in samples")

def run_dealplus(input_h5ad, out_h5ad, out_umap, other1_key, other2_key):
    sc.set_figure_params(dpi_save=300, frameon=False, figsize=(10, 6))
    adata = sc.read_h5ad(input_h5ad)
    #add biosample.celltype column
    adata.obs['biosample.celltype'] = adata.obs['biosample'].astype(str) + '_' + adata.obs['celltype'].astype(str)
    print(adata.obs["biosample.celltype"].unique())
    #visual
    with PdfPages(out_umap) as pdf:
        sc.pl.umap(adata, color=["biosample", "celltype", "biosample.celltype"], legend_loc='on data', ncols=1)
        plt.savefig(pdf, format='pdf', dpi=300, bbox_inches='tight')
        plt.close()
        required_cols = ['total_counts', 'n_genes']
        if all(col in adata.obs.columns for col in required_cols):
            sc.pl.violin(adata, keys=['total_counts'], log=False, groupby="biosample", show=False)
            plt.savefig(pdf, format='pdf', dpi=300, bbox_inches='tight')
            plt.close()
            sc.pl.violin(adata, keys=['n_genes'], log=False, groupby="biosample", show=False)
            plt.savefig(pdf, format='pdf', dpi=300, bbox_inches='tight')
            plt.close()
        else:
            print("obs lacked total_counts or n_genes column")
    # 检查 obs 中是否存在这两个列
    if other1_key in adata.obs.columns and other2_key in adata.obs.columns:
        # 检查是否满足条件：other1_key 不等于 "biosample" 或 other2_key 不等于 "celltype"
        if other1_key != "biosample" or other2_key != "celltype":
            # 添加组合列
            adata.obs['other1_other2'] = adata.obs[other1_key].astype(str) + '_' + adata.obs[other2_key].astype(str)

            # 打印组合列的唯一值
            print(adata.obs["other1_other2"].unique())

            # 绘制 UMAP 并保存为 PDF
            with PdfPages('other'+out_umap) as pdf:
                sc.pl.umap(adata, color=[other1_key, other2_key, "other1_other2"], legend_loc='on data', ncols=1, show=False)
                plt.savefig(pdf, format='pdf', dpi=300, bbox_inches='tight')
                plt.close()
        else:
            print("other1_key 或 other2_key 为默认值，跳过 UMAP 绘图。")
    else:
        print("obs 中缺少指定的列，跳过 UMAP 绘图。")
    
    click.echo("Save output")
    adata.write(filename=out_h5ad,compression="gzip")


if __name__ == '__main__':
    run_dealplus()
